userland/Makefile

```diff
@@ -359,58 +67,51 @@ $(BOARD_BUILDDIR)/app.elf: $(OBJS) $(TOCK_USERLAND_BASE_DIR)/newlib/libc.a $(LIB
+	$(Q)$(READELF) -p .GCC.command.line $$< 2>&1 | grep -q "does not exist" && { echo "Error: Missing section .GCC.command.line"; echo ""; echo "Tock requires that applications are built with"; echo "  -frecord-gcc-switches"; echo "to validate that all required flags were used"; echo ""; echo "You can skip this check by defining the make variable TOCK_NO_CHECK_SWITCHES"; exit 1; } || exit 0
+	$(Q)$(READELF) -p .GCC.command.line $$< | grep -q -- -msingle-pic-base && $(READELF) -p .GCC.command.line $$< | grep -q -- -mpic-register=r9 && $(READELF) -p .GCC.command.line $$< | grep -q -- -mno-pic-data-is-text-relative || { echo "Error: Missing required build flags."; echo ""; echo "Tock requires applications are built with"; echo "  -msingle-pic-base"; echo "  -mpic-register=r9"; echo "  -mno-pic-data-is-text-relative"; echo "But one or more of these flags are missing"; echo ""; echo "To see the flags your application was built with, run"; echo "$(READELF) -p .GCC.command.line $$<"; echo ""; exit 1; }
 endif
 
-.PHONY:
-clean::
-	rm -Rf $(BUILDDIR)
-	rm -Rf $(BOARD_BUILDDIR)
+OBJS_NO_ARCHIVES += $(filter %.o,$(patsubst %.c,$(BUILDDIR)/$(1)/%.o,$(C_SRCS)) $(patsubst %.cc,$(BUILDDIR)/$(1)/%.o,$(filter %.cc, $(CXX_SRCS))) $(patsubst %.cpp,$(BUILDDIR)/$(1)/%.o,$(filter %.cpp, $(CXX_SRCS))))
+endef
+$(foreach arch, $(TOCK_ARCHS), $(eval $(call BUILD_RULES,$(arch))))
 
 
+# Rules for building apps
 
-#########################################################################################
-## Pretty-printing rules
+.PHONY:	all
+all:	$(foreach arch, $(TOCK_ARCHS), $(BUILDDIR)/$(arch)/$(arch).bin) size tab
 
-# If environment variable V is non-empty, be verbose
-ifneq ($(V),)
-Q=
-TRACE_BIN =
-TRACE_DEP =
-TRACE_CC  =
-TRACE_CXX =
-TRACE_LD  =
-TRACE_AR  =
-TRACE_AS  =
-TRACE_LST =
-ELF2TBF_ARGS += -v
-else
-Q=@
-TRACE_BIN = @echo " BIN       " $@
-TRACE_DEP = @echo " DEP       " $<
-TRACE_CC  = @echo "  CC       " $<
-TRACE_CXX = @echo " CXX       " $<
-TRACE_LD  = @echo "  LD       " $@
-TRACE_AR  = @echo "  AR       " $@
-TRACE_AS  = @echo "  AS       " $<
-TRACE_LST = @echo " LST       " $<
-endif
+.PHONY: size
+size:	$(foreach arch, $(TOCK_ARCHS), $(BUILDDIR)/$(arch)/$(arch).elf)
+	@$(SIZE) $^
+
+.PHONY: debug
+debug:	$(foreach arch, $(TOCK_ARCHS), $(BUILDDIR)/$(arch)/$(arch).lst)
+
+.PHONY:
+clean::
+	rm -Rf $(BUILDDIR)
 
+.PHONY:
+tab: $(foreach arch, $(TOCK_ARCHS), $(BUILDDIR)/$(arch)/$(arch).bin)
+	$(TOCK_USERLAND_BASE_DIR)/tools/tab/create_tab.py $(BUILDDIR)/$(PACKAGE_NAME).tab $(PACKAGE_NAME) $^
```

Comment:
Well, it's PHONY because it's not pointing to a particular file. It really ought to be something like:

```Makefile
.PHONY: tab
tab: path/to/tab/fille.tab

path/to/tab/fille.tab: $(foreach arch, $(TOCK_ARCHS), $(BUILDDIR)/$(arch)/$(arch).bin)
   ...
```